langgraph/checkpoint/sqlite.py

```diff
@@ -382,3 +434,39 @@ def put(
                 "thread_ts": checkpoint["ts"],
             }
         }
+
+
+def search_where(metadata_query: CheckpointMetadata) -> str:
+    """Return WHERE clause for (a)search() given metadata query.
+
+    This method returns the operator as well (=, IS).
+    """
+
+    def _where_value(query_value: Any) -> str:
+        if query_value is None:
+            return "IS NULL"
+        elif isinstance(query_value, str):
+            return f"= '{query_value}'"
+        elif isinstance(query_value, int) or isinstance(query_value, float):
+            return f"= {query_value}"
+        elif isinstance(query_value, bool):
+            return f"= {1 if query_value else 0}"
+        elif isinstance(query_value, dict) or isinstance(query_value, list):
+            # query value for JSON object cannot have trailing space after separators (, :)
+            # SQLite json_extract() returns JSON string without whitespace
+            return f"= '{json.dumps(query_value, separators=(',', ':'))}'"
+        else:
+            return f"= '{str(query_value)}'"
+
+    where = "WHERE "
+    for query_key, query_value in metadata_query.items():
+        where += f"json_extract(CAST(metadata AS TEXT), '$.{query_key}') {_where_value(query_value)} AND "
```

Comment:
Not necessarily performant to call `json_extract()` for each metadata query key, but it's the simplest way to do it without creating a subtable, etc.